app/roles/api.lua

```diff
@@ -0,0 +1,159 @@
+local cartridge = require('cartridge')
+local errors = require('errors')
+local http_client = require('http.client')
+local json = require('json')
+local log = require('log')
+
+local err_vshard_router = errors.new_class('Vshard routing error')
+local err_http_client = errors.new_class('HTTP client error')
+
+-- делаем запрос в тарантул и пытаемся получить данные
+-- если данные есть:
+--      отдаем их клиенту
+-- если данных нет:
+--      делаем запрос к сервису прогноза погоды
+--      сохраняем в тарантул
+--      отдаем клиенту
+local function http_weather_get_handler(req)
+    local query = req.query
+
+    local latitude = req:query_param('latitude')
+    local longitude = req:query_param('longitude')
+
+    -- if long or lat empty
+    if not longitude or not latitude then
+        -- return client error
+        return bad_response(req, 'longitude or latitude error', 400)
+    end
+
+    longitude = tonumber(longitude)
+    latitude = tonumber(latitude)
+
+    -- get router and create bucket id
+    local router = cartridge.service_get('vshard-router').get()
+    local bucket_by_data = calc_bucket_by_coords(longitude, latitude)
+    local bucket_id = router:bucket_id(bucket_by_data)
+
+    -- trying to get weather forecast from cache (taranrool)
+    local t_resp, err = err_vshard_router:pcall(
+        router.call,
+        router,
+        bucket_id,
+        'read',
+        'weather_forecast_get',
+        {longitude, latitude}
+    )
+    -- tarantool request error
+    if err then
+        log.error(err.message)
+
+        return bad_response(req, err.message, 500)
+    end
+
+    -- successfully retrieved data from cache (taranrool)
+    if t_resp then
+        return ok_response(req, json.decode(t_resp.raw_data))
+    end
+
+    -- make request to weather service
+    local raw_forecast, err = err_http_client:pcall(
+        make_weather_query,
+        longitude,
+        latitude
+    )
+    if err then
+        log.error(err.message)
+
+        return bad_response(req, err.message, 500)
+    end
+
+    local forecast = json.decode(raw_forecast)
+    forecast.raw_data = raw_forecast
+    forecast.bucket_id = bucket_id
+    -- HACK: weather service sometimes respond slightly another coords
+    -- we use requested coords for save into cache (tarantool)
+    forecast.latitude = latitude
+    forecast.longitude = longitude
+
+    -- trying to save into cache (tarantool)
+    local t_resp, err = err_vshard_router:pcall(
+        router.call,
+        router,
+        bucket_id,
+        'write',
+        'weather_forecast_add',
+        {forecast}
+    )
+    if err then
```

Comment:
not sure if we should fail in case of such problems - we are still capable of requesting the source (weather API)

Reply:
the same applies to cache update

Reply:
fixed